### PR TITLE
schema cache syncing without relying on a Postgres table

### PR DESCRIPTION
**Pull request by rakeshkky**
### Description

This is an incremental PR towards #5797. It avoids using `hdb_cataog.hdb_schema_update` table and associated trigger to notify schema syncing events to other server instances. Instead it is achieved in a much simpler way by directly using `pg_notify` procedure to publish event payload.

Other things this PR does:-
- Now `initialiseCtx` function is only used in `serve` command
- Using new `GlobalCtx` type for essentials required for all graphql-engine commands

### Changelog

- [ ] `CHANGELOG.md` is updated with user-facing content relevant to this PR. If no changelog is required, then add the `no-changelog-required` label.

### Affected components

- [x] Server

#### Catalog upgrade

Does this PR change Hasura Catalog version?
- [x] No
- [ ] Yes

#### Metadata

Does this PR add a new Metadata feature?
- [x] No
- [ ] Yes

#### Breaking changes

- [x] No Breaking changes
- [ ] There are breaking changes

### CODE REVIEW

File: server/src-lib/Hasura/App.hs [hunk 1]
```diff
@@ -156,6 +155,23 @@ mkPGLogger :: Logger Hasura -> Q.PGLogger
 mkPGLogger (Logger logger) (Q.PLERetryMsg msg) =
   logger $ PGLog LevelWarn msg
 
+-- | Context required for all graphql-engine commands
```
Comment (abooij) [inline server/src-lib/Hasura/App.hs:158]: What does "all graphql-engine commands" mean? Does this refer to CLI commands? graphql-engine supports a variety of commands through a variety of interfaces and protocols :-)

File: server/src-lib/Hasura/App.hs [hunk 2]
```diff
@@ -156,6 +155,23 @@ mkPGLogger :: Logger Hasura -> Q.PGLogger
...
+data GlobalCtx
+  = GlobalCtx
+  { _gcHttpManager :: !HTTP.Manager
+  , _gcInstanceId  :: !InstanceId
```
Comment (abooij) [inline server/src-lib/Hasura/App.hs:162]: `_gcInstanceId` is only used by `serve`, right? And `InitCtx` is also only used by `serve`, right? If these two are right, then perhaps we can move `_gcInstanceId` to `InitCtx`, and rename `InitCtx` to something more fitting, such as `ServeCtx`.

File: server/src-lib/Hasura/App.hs [hunk 3]
```diff
@@ -181,71 +198,64 @@ data Loggers
 newtype AppM a = AppM { unAppM :: IO a }
   deriving (Functor, Applicative, Monad, MonadIO, MonadBase IO, MonadBaseControl IO, MonadCatch, MonadThrow, MonadMask)
 
--- | this function initializes the catalog and returns an @InitCtx@, based on the command given
--- - for serve command it creates a proper PG connection pool
--- - for other commands, it creates a minimal pool
--- this exists as a separate function because the context (logger, http manager, pg pool) can be
--- used by other functions as well
+-- | Initializes the catalog and returns an @InitCtx@ for given @'ServeOptions'
```
Comment (abooij) [inline server/src-lib/Hasura/App.hs:201]: As a rule of thumb: a comment that does little more than spell out the Haskell type of the thing being documented is not documentation. If the comments contain information that can be extracted from the Haskell code itself, they're not adding information, just duplicating it.

[Here](https://carlanderson.xyz/how-to-write-useful-comments/)'s one of many guides you might find useful to read about writing comments.

**Review by abooij — COMMENTED** [review]
I can't reliably judge the syncing logic itself. It seems to me that the core of tho changes are in `fetchLastUpdate`, `recordSchemaUpdate`, and `notifySchemaCacheSync`. With the old `fetchLastUpdate` and `recordSchemaUpdate`, I think I understood the logic: `record` sends a message to all other graphql-engine instances, and `fetch` received messages. Is `notify` the new `record` or the new `fetch`? I believe it's the former. If so, where's the new `fetch`? Where's the code that receives these PG notifications?

Edit: Sorry, perhaps the failing PR was just a flake. [This used to say: @rakeshkky, please invite reviewers *after* you got the PR to pass CI. Otherwise, you ensure that the reviewer does *not* review the code that will end up getting merged.]

File: server/src-lib/Hasura/App.hs (hunk 1 above)
Comment (rakeshkky) [reply server/src-lib/Hasura/App.hs:158]: > Does this refer to CLI commands? 

Yes.

Comment (abooij) [reply server/src-lib/Hasura/App.hs:158]: Could you please clarify this in the comment?

**Comment by rakeshkky** [comment]
@0x777 Please take a look at this PR. You have the better context of the schema syncing process than other reviewers.

**Pull request closed without merging**